lib/client.js

```diff
@@ -1,427 +1,427 @@
-
-            if (connected) {
-                adapter.log.info('Disconnected from ' + config.url);
-                connected = false;
-                adapter.setState('info.connection', connected, true);
-            }
-        });
-
-        client.on('close', err => {
-            if (connected) {
-                adapter.log.info('Disconnected from ' + config.url + ': ' + err);
-                connected = false;
-                adapter.setState('info.connection', connected, true);
-            }
-        });
-    })(adapter.config);
-
-    process.on('uncaughtException', err => adapter.log.error('uncaughtException: ' + err));
-
-    return this;
-}
-
-module.exports = MQTTClient;
+'use strict';
+
+const mqtt            = require('mqtt');
+const utils           = require('@iobroker/adapter-core');
+const tools           = require(utils.controllerDir + '/lib/tools');
+const state2string    = require(__dirname + '/common').state2string;
+const convertTopic2id = require(__dirname + '/common').convertTopic2id;
+const convertID2topic = require(__dirname + '/common').convertID2topic;
+
+const messageboxRegex = new RegExp('\\.messagebox$');
+
+function getAppName() {
+    const parts = __dirname.replace(/\\/g, '/').split('/');
+    return parts[parts.length - 2].split('.')[0];
+}
+utils.appName = getAppName();
+
+function MQTTClient(adapter, states) {
+    if (!(this instanceof MQTTClient)) return new MQTTClient(adapter, states);
+
+    let client    = null;
+    const topic2id  = {};
+    const id2topic  = {};
+    const namespaceRegEx = new RegExp('^' + adapter.namespace.replace('.', '\\.') + '\\.');
+    let connected = false;
+
+    this.destroy = () => {
+        if (client) {
+            client.end();
+            client = null;
+        }
+    };
+
+    this.onStateChange = (id, state) => send2Server(id, state);
+    this.onMessage = (topic, message) => publishMessage(topic, message);
+
+    function send2Server(id, state, cb) {
+        if (!client) return;
+        const topic = id2topic[id];
+        adapter.log.info('send2Server ' + id + '[' + topic + ']');
+        if (!topic) {
+            adapter.getForeignObject(id, (err, obj) => {
+                if (!client) return;
+                if (!obj) {
+                    adapter.log.warn('Cannot resolve topic name for ID: ' + id + ' (object not found)');
+                    if (cb) cb(id);
+                    return;
+                } else if (!obj.native || !obj.native.topic) {
+                    id2topic[obj._id] = convertID2topic(obj._id, null, adapter.config.prefix, adapter.namespace);
+                } else {
+                    id2topic[obj._id] = obj.native.topic;
+                }
+                send2Server(obj._id, state, cb);
+            });
+            return;
+        }
+
+        publishMessage(
+            adapter.config.extraSet && state && !state.ack ? topic + '/set' : topic,
+            state ? state : null

```

Comment:
Why no state2string(state.val) anymore?

Reply:
The idea of the pull request is that the message to sent is the complete JSON object as a string. In that line the function publisMessage is called with only the atomar value of "state.val". That means the mqtt client in line 74 can't decide to send either the atomar vlaue of "state.val" or the complete JSON object "state". I would do the decision to send eitehr the atomar value or the JOSN object at the latest call (here "state2String()").

Additionally it would make maybe sense to use the function "publishMessage" at line 104 and 110 as well. Then the mqtt library function "client.publish" is used only once. What do you think?

Reply:
Ok